File: src/main/java/gargoyle/ct/resource/internal/ClasspathResource.java
package gargoyle.ct.resource.internal;

import gargoyle.ct.resource.Resource;

import java.io.IOException;
import java.net.URL;

/**
 * represents a {@link Resource} withing application class path (ex. in jar file). writing is mainly unsupported
 */
public class ClasspathResource extends VirtualResource {
    final ClassLoader loader;

    public ClasspathResource(String location) {
        this(ClasspathResource.class.getClassLoader(), location);
    }

    public ClasspathResource(ClassLoader loader, String location) {
        super(location);
        this.loader = loader;
    }

    ClasspathResource(ClassLoader loader, Resource base, String location) {
        super(base, location);
        this.loader = loader;
    }

    @Override
    protected ClasspathResource createResource(Resource base, String location) {
        return base == null ?
                new ClasspathResource(loader, location) :
                new ClasspathResource(getLoader(base), base, location);
    }

    @Override
    public URL toURL() throws IOException {
        return loader.getResource(getLocation());
    }

    ClassLoader getLoader(Resource resource) {
        if (resource instanceof ClasspathResource) {
            return ((ClasspathResource) resource).loader;
        }
        return loader;
    }

    @Override
    public boolean exists() {
        try {
            return exists(toURL());
        } catch (IOException ex) {
            return false;
        }
    }
}

File: src/main/kotlin/gargoyle/ct/resource/internal/ClasspathResource.kt
package gargoyle.ct.resource.internal

import gargoyle.ct.resource.Resource
import java.io.IOException
import java.net.URL

/**
 * represents a [Resource] withing application class path (ex. in jar file). writing is mainly unsupported
 */
open class ClasspathResource : VirtualResource<ClasspathResource> {
    val loader: ClassLoader

    constructor(location: String) : this(ClasspathResource::class.java.classLoader, location)
    constructor(loader: ClassLoader, location: String) : super(location) {
        this.loader = loader
    }

    internal constructor(loader: ClassLoader, base: ClasspathResource, location: String) : super(base, location) {
        this.loader = loader
    }

    override fun createResource(
        base: VirtualResource<ClasspathResource>?,
        location: String
    ): VirtualResource<ClasspathResource> {
        return if (base == null) ClasspathResource(loader, location) else ClasspathResource(
            getLoader(base),
            base as ClasspathResource,
            location
        )
    }

    @Throws(IOException::class)
    override fun toURL(): URL {
        return loader.getResource(location)!!
    }

    fun getLoader(resource: Resource?): ClassLoader {
        return if (resource is ClasspathResource) resource.loader else loader
    }

    override fun exists(): Boolean {
        return try {
            exists(toURL())
        } catch (ex: IOException) {
            false
        }
    }
}
